python/ray/dag/compiled_dag_node.py

```diff
@@ -334,34 +361,52 @@ def _get_or_compile(
             elif isinstance(task.dag_node, InputNode):
                 task.output_channel = Channel(
                     buffer_size_bytes=self._buffer_size_bytes,
-                    num_readers=task.num_readers,
+                    num_readers=task.num_readers if task.num_readers else 1,
                 )
             else:
-                assert isinstance(task.dag_node, MultiOutputNode)
+                assert isinstance(task.dag_node, MultiOutputNode) or isinstance(
+                    task.dag_node, InputAttributeNode
+                )
 
             for idx in task.downstream_node_idxs:
                 queue.append(idx)
 
         for node_idx, task in self.idx_to_task.items():
-            if node_idx == self.input_task_idx:
-                # We don't need to assign an actual task for the input node.
+            if node_idx in self.input_task_idxs:
+                # We don't need to assign an actual task for the input node(s).
                 continue
 
             if node_idx == self.output_task_idx:
-                # We don't need to assign an actual task for the input node.
+                # We don't need to assign an actual task for the output node.
                 continue
 
             resolved_args = []
+            # for channel args, the index of read result to take, -1 stands for
+            # taking the whole input
```

Comment:
If I understand correctly, this is a case like

inp = InputNode()
f.remote(inp.x)
g.remote(inp.y)

-> f and g both read {x, y} from the same channel, but read different indices

Could you add a comment explaining this? It's a bit unclear - there is an alternative design that would separate x and y into two separate channels.

Reply:
thx for pointing out! added this example to comments